### PR TITLE
go/store/prolly: 3-way Merge for Prolly Trees

### PR DESCRIPTION
**Pull request by andrew-wm-arthur**

### CODE REVIEW

File: go/store/prolly/utils_test.go [hunk 1]
```diff
@@ -77,23 +77,34 @@ func randomTuplePairs(count int, keyDesc, valDesc val.TupleDesc) (items [][2]val
...
+			break
+		}
+
+		// replace duplicates and validate again
+		for _, d := range dupes {
+			items[d][0] = randomTuple(keyBuilder)
 		}
+		dupes = dupes[:0]
 	}
 	return
```
Comment (bheni) [inline go/store/prolly/utils_test.go:101]: not a fan of naked returns.  Might `return items` while you're here

File: go/store/prolly/tree_merge.go [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+		return Map{}, err
+	}
+
+	eg, ctx := errgroup.WithContext(ctx)
+	buf := newPatchBuffer(patchBufferSize)
+
+	// iterate |ld| and |rd| in parallel, populating |buf|
+	eg.Go(func() (err error) {
+		defer func() {
+			if cerr := buf.close(); err != nil {
```
Comment (bheni) [inline go/store/prolly/tree_merge.go:57]: I think this should be if `err == nil` so it only overwrites the error when there is no error already existing.  This currently overwrites the error only if there is a pre-existing error.

File: go/store/prolly/tree_merge.go [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+					err = buf.sendPatch(ctx, resolved)
+				}
+				if err != nil {
+					return err
+				}
+			}
+
+			left, err = l.Next(ctx)
+			if err == io.EOF {
+				err, lok = nil, false
```
Comment (bheni) [inline go/store/prolly/tree_merge.go:182]: wow, such line savings
much concise

**Review by bheni — APPROVED** [review]
lgtm.  One bug. few comments/suggestions.

File: go/store/prolly/map_test.go [hunk 4]
```diff
@@ -20,14 +20,15 @@ import (
...
+	"time"
 
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 
 	"github.com/dolthub/dolt/go/store/val"
 )
 
-var testRand = rand.New(rand.NewSource(1))
+var testRand = rand.New(rand.NewSource(time.Now().UnixNano()))
```
[the remark this answers is not in the document]
Comment (andrew-wm-arthur) [reply go/store/prolly/map_test.go:31]: For now I'm going to revert this back to a constant seed. I'll take another pass at this to randomize test data

File: go/store/prolly/tree_merge.go (hunk 2 above)
Comment (andrew-wm-arthur) [reply go/store/prolly/tree_merge.go:57]: dangerous!

**Pull request merged**